src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala

```diff
@@ -17,86 +15,120 @@ import scala.concurrent.{ExecutionContext, Future}
+  def listUserResourceActions(resource: Resource, userInfo: UserInfo): Future[Set[ResourceAction]] = {
+    listResourceAccessPoliciesForUser(resource, userInfo).map { matchingPolicies =>
       matchingPolicies.flatMap(_.actions)
     }
   }
 
-  def listUserResourceRoles(resourceTypeName: ResourceTypeName, resourceName: ResourceName, userInfo: UserInfo): Future[Set[ResourceRoleName]] = {
-    listResourceAccessPoliciesForUser(resourceTypeName, resourceName, userInfo).map { matchingPolicies =>
-      matchingPolicies.flatMap(_.role)
+  def listUserResourceRoles(resource: Resource, userInfo: UserInfo): Future[Set[ResourceRoleName]] = {
+    listResourceAccessPoliciesForUser(resource, userInfo).map { matchingPolicies =>
+      matchingPolicies.flatMap(_.roles)
     }
   }
 
-  private def listResourceAccessPoliciesForUser(resourceTypeName: ResourceTypeName, resourceName: ResourceName, userInfo: UserInfo): Future[Set[AccessPolicy]] = {
-    for {
-      policies <- accessPolicyDAO.listAccessPolicies(resourceTypeName, resourceName)
-      groups <- directoryDAO.listUsersGroups(userInfo.userId)
-    } yield {
-      policies.filter { policy =>
-        policy.subject match {
-          case user: WorkbenchUserId => userInfo.userId == user
-          case group: WorkbenchGroupName => groups.contains(group)
+  private def listResourceAccessPoliciesForUser(resource: Resource, userInfo: UserInfo): Future[Set[AccessPolicy]] = {
+    accessPolicyDAO.listAccessPoliciesForUser(resource, userInfo.userId)
+  }
+
+  //Overwrites an existing policy (keyed by resourceType/resourceId/policyName), saves a new one if it doesn't exist yet
+  def overwritePolicy(resourceType: ResourceType, policyName: String, resource: Resource, policyMembership: AccessPolicyMembership, userInfo: UserInfo): Future[AccessPolicy] = {
+    requireAction(resource, SamResourceActions.alterPolicies, userInfo) {
+      if(!policyMembership.actions.subsetOf(resourceType.actions))
+        throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"You have specified an invalid action for resource type ${resourceType.name}. Valid actions are: ${resourceType.actions.mkString(", ")}"))
+      if(!policyMembership.roles.subsetOf(resourceType.roles.map(_.roleName)))
+        throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"You have specified an invalid role for resource type ${resourceType.name}. Valid roles are: ${resourceType.roles.map(_.roleName).mkString(", ")}"))
+
+      val subjectsFromEmails = Future.traverse(policyMembership.memberEmails) {
+        directoryDAO.loadSubjectFromEmail
+      }.map(_.flatten)
+
+      val email = toGoogleGroupEmail(policyName, resource)
+
+      val actionsByRole = resourceType.roles.map(r => r.roleName -> r.actions).toMap
+      val impliedActionsFromRoles = policyMembership.roles.flatMap(actionsByRole)
```

Comment:
this solution works nicely if we don't expect roles to change. if we do, it's back to the drawing board.

potential other solution is, when loading access policies and there is an attached role, do a lookup to find the policy for that role and tack on its actions (instead of storing them as attrs in opendj)